services/py-api/src/service/hackathon_service.py

```diff
@@ -441,3 +443,85 @@ def send_successful_registration_email(
             return err
 
         return None
+
+    async def retrieve_and_categorize_random_participants(
+        self,
+    ) -> Result[Tuple[List[Participant], List[Participant]], Exception]:
+        programming_oriented = []
+        non_programming_oriented = []
+        # Fetch all the verified random participants
+        result = await self._participant_repo.get_verified_random_participants()
+
+        # Return the result to the upper layer in case of an Exception
+        if is_err(result):
+            return result
+
+        # Group the into categories programming oriented, non-programming oriented
+        for participant in result.ok_value:
+            if participant.programming_level == "I am not participating as a programmer":
+                non_programming_oriented.append(participant)
+            else:
+                programming_oriented.append(participant)
+
+        return Ok((programming_oriented, non_programming_oriented))
+
+    def form_random_participant_teams(
+        self, prog_participants: list[Participant], non_prog_participants: list[Participant]
+    ) -> List[RandomTeam]:
+        # Calculate the number of hackathon participants
+        number_of_random_participants = len(prog_participants) + len(non_prog_participants)
+        # Calculate the number of teams that are going to be needed for the given number of participants
+        number_of_teams = ceil(number_of_random_participants / self.MAX_NUMBER_OF_TEAM_MEMBERS)
+
+        # Create a dictionary that is going to hold the participants for each team
+        teams = []
+        # Populate the dictionary with the Random Teams named `RandomTeam{i}`
+        for i in range(number_of_teams):
+            teams.append(RandomTeam(team=Team(name=f"RandomTeam{i}", is_verified=True), participants=[]))
+
+        # Spread all the programming experienced participants to the teams in a Round Robin (RR) manner
+        ctr = 0  # initialize a counter
+        while len(prog_participants) > 0:
+            teams[ctr % number_of_teams]["participants"].append(prog_participants.pop())
```

Comment:
maybe add a comment. If I am understanding this correctyl:

1. Until the list of prog_participants is empty
get the Dict at index (idk why this index is calculated like that, myabe have to look up RR impl). From it get the other list for key participants and append to it a prog_participant (removing it from the list above)